File: frontend/src/lib/monitoring.ts
"use client";

type MonitoringContext = {
  path?: string;
  method?: string;
  status?: number;
  feature?: string;
  expected?: boolean;
  retryAfterSeconds?: number | null;
};

type SentryScope = {
  setTag: (key: string, value: string) => void;
  setLevel: (level: "warning" | "error") => void;
  setContext: (name: string, context: Record<string, unknown>) => void;
};

type SentryClient = {
  init: (config: Record<string, unknown>) => void;
  withScope: (callback: (scope: SentryScope) => void) => void;
  captureException: (error: unknown) => void;
};

declare global {
  interface Window {
    Sentry?: SentryClient;
    __internrouteSentryInitialized?: boolean;
  }
}

const SENTRY_DSN = (process.env.NEXT_PUBLIC_SENTRY_DSN || "").trim();
const SENTRY_ENVIRONMENT = (process.env.NEXT_PUBLIC_SENTRY_ENVIRONMENT || process.env.NODE_ENV || "development").trim();
const SENTRY_RELEASE = (process.env.NEXT_PUBLIC_SENTRY_RELEASE || "").trim() || undefined;
const SENTRY_TRACES_SAMPLE_RATE = Number.parseFloat(
  (process.env.NEXT_PUBLIC_SENTRY_TRACES_SAMPLE_RATE || "0").trim()
);

function _isSentryEnabled(): boolean {
  return SENTRY_DSN.length > 0;
}

function _sentryClient(): SentryClient | null {
  if (typeof window === "undefined") return null;
  return window.Sentry || null;
}

function _sanitizeEvent(event: Record<string, unknown>): Record<string, unknown> | null {
  const tags = (event.tags || {}) as Record<string, unknown>;
  if (String(tags.expected || "") === "true") {
    return null;
  }
  const request = event.request as Record<string, unknown> | undefined;
  if (request && typeof request === "object") {
    delete request.data;
    delete request.cookies;
    const headers = request.headers as Record<string, unknown> | undefined;
    if (headers && typeof headers === "object") {
      for (const key of Object.keys(headers)) {
        const lowered = key.toLowerCase();
        if (lowered === "authorization" || lowered === "cookie" || lowered === "set-cookie" || lowered === "x-csrf-token") {
          headers[key] = "[Filtered]";
        }
      }
    }
  }
  return event;
}

export function initFrontendMonitoring(): boolean {
  if (!_isSentryEnabled() || typeof window === "undefined") {
    return false;
  }
  if (window.__internrouteSentryInitialized) {
    return true;
  }
  const client = _sentryClient();
  if (!client) {
    return false;
  }

  const tracesSampleRate = Number.isFinite(SENTRY_TRACES_SAMPLE_RATE)
    ? Math.max(0, Math.min(1, SENTRY_TRACES_SAMPLE_RATE))
    : 0;

  client.init({
    dsn: SENTRY_DSN,
    environment: SENTRY_ENVIRONMENT,
    release: SENTRY_RELEASE,
    tracesSampleRate,
    sendDefaultPii: false,
    beforeSend: _sanitizeEvent,
  });
  window.__internrouteSentryInitialized = true;
  return true;
}

export function captureFrontendError(error: unknown, context: MonitoringContext = {}): void {
  if (context.expected) return;
  if (context.path === "/auth/me" && context.status === 401) return;
  if (typeof context.status === "number" && context.status < 500) return;
  if (!_isSentryEnabled() || typeof window === "undefined") return;

  if (!window.__internrouteSentryInitialized) {
    initFrontendMonitoring();
  }
  const client = _sentryClient();
  if (!client) return;

  const wrappedError = error instanceof Error ? error : new Error(String(error));

  client.withScope((scope) => {
    scope.setTag("feature", context.feature || "frontend");
    if (context.path) scope.setTag("path", context.path);
    if (context.method) scope.setTag("method", context.method);
    if (typeof context.status === "number") scope.setTag("http_status", String(context.status));
    scope.setLevel(typeof context.status === "number" && context.status >= 500 ? "error" : "warning");
    scope.setContext("api", {
      path: context.path || null,
      method: context.method || null,
      status: typeof context.status === "number" ? context.status : null,
      retry_after_seconds: context.retryAfterSeconds ?? null,
    });
    client.captureException(wrappedError);
  });
}

File: backend/app/monitoring.py
from __future__ import annotations

import logging
from typing import Any

from flask import Flask

_SENTRY_INITIALIZED = False

try:
  import sentry_sdk
  from sentry_sdk.integrations.flask import FlaskIntegration
  from sentry_sdk.integrations.logging import LoggingIntegration
  from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
except Exception:  # pragma: no cover
  sentry_sdk = None
  FlaskIntegration = None
  LoggingIntegration = None
  SqlalchemyIntegration = None


def _before_send(event: dict[str, Any], hint: dict[str, Any]):
  request = event.get("request")
  if isinstance(request, dict):
    headers = request.get("headers")
    if isinstance(headers, dict):
      for key in list(headers.keys()):
        if key.lower() in {"authorization", "cookie", "set-cookie", "x-csrf-token"}:
          headers[key] = "[Filtered]"
    request.pop("data", None)
    request.pop("cookies", None)

  user = event.get("user")
  if isinstance(user, dict):
    user.pop("email", None)
    user.pop("ip_address", None)
    user.pop("username", None)

  return event


def init_error_monitoring(app: Flask) -> None:
  global _SENTRY_INITIALIZED
  dsn = (app.config.get("SENTRY_DSN") or "").strip()
  if not dsn:
    app.logger.info("Sentry monitoring disabled: SENTRY_DSN is not configured.")
    return

  if sentry_sdk is None or FlaskIntegration is None or LoggingIntegration is None or SqlalchemyIntegration is None:
    app.logger.warning("Sentry DSN is configured but sentry-sdk is not installed.")
    return

  sentry_sdk.init(
    dsn=dsn,
    environment=app.config.get("SENTRY_ENVIRONMENT"),
    release=app.config.get("SENTRY_RELEASE"),
    send_default_pii=False,
    traces_sample_rate=float(app.config.get("SENTRY_TRACES_SAMPLE_RATE", 0.0)),
    integrations=[
      FlaskIntegration(transaction_style="endpoint"),
      SqlalchemyIntegration(),
      LoggingIntegration(level=logging.INFO, event_level=logging.ERROR),
    ],
    before_send=_before_send,
  )
  _SENTRY_INITIALIZED = True
  app.logger.info("Sentry monitoring enabled.")


def capture_monitored_exception(
  err: Exception,
  *,
  tags: dict[str, str] | None = None,
  context: dict[str, Any] | None = None,
) -> None:
  if not _SENTRY_INITIALIZED or sentry_sdk is None:
    return

  with sentry_sdk.push_scope() as scope:
    for key, value in (tags or {}).items():
      scope.set_tag(str(key), str(value))
    if context:
      scope.set_context("app_context", context)
    sentry_sdk.capture_exception(err)
